### PR TITLE
Show Language Server Logs

### PR DESCRIPTION
**Pull request by somebody1234**
### Pull Request Description
- Frontend side of https://github.com/enso-org/cloud-v2/issues/1046
  - Add endpoint to list project sessions
  - Add button to project sessions to view logs associated with a specific project session

### Important Notes
None

### Testing instructions
Just make sure that:
- the new project sessions tab works (and does not show up for non-projects)
- logs show up as expected.

### Screenshots

Project sessions list:
image

Logs modal:
image

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] ~~The documentation has been updated, if necessary.~~
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] ~~Unit tests have been written where possible.~~
  - [x] ~~If GUI codebase was changed, the GUI was tested when built using `./run ide build`.~~

### CODE REVIEW

File: app/ide-desktop/lib/dashboard/src/modals/LogsModal.tsx [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  React.useEffect(() => {
+    if (logsRaw instanceof Promise) {
+      void logsRaw.then(newLogs => {
+        setLogs(newLogs)
+      })
+    }
+  }, [logsRaw])
+
+  return (
+    <Modal centered className="bg-dim">
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/modals/LogsModal.tsx]: Can we use `ariaComponents.Dialog type="fullscreen"` here?

File: app/ide-desktop/lib/dashboard/src/modals/LogsModal.tsx [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+/** A modal for showing logs for a project. */
+export default function LogsModal(props: LogsModalProps) {
+  const { logs: logsRaw } = props
+  const [logs, setLogs] = React.useState(() => (logsRaw instanceof Promise ? null : logsRaw))
+  const { getText } = textProvider.useText()
+
+  React.useEffect(() => {
+    if (logsRaw instanceof Promise) {
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/modals/LogsModal.tsx]: Use please `useQuery` here fetching is `useEffect` is a bad practice

File: app/ide-desktop/lib/dashboard/src/modals/LogsModal.tsx [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        className="pointer-events-auto relative flex h-4/5 w-4/5 flex-col gap-4 overflow-auto rounded-default p-4 before:absolute before:inset before:h-full before:w-full before:rounded-default before:bg-selected-frame before:backdrop-blur-3xl"
+        onClick={() => {
+          event?.stopPropagation()
+        }}
+      >
+        <aria.Heading level={2} className="relative text-lg font-bold">
+          {getText('logs')}
+        </aria.Heading>
+        <pre className="relative overflow-auto whitespace-pre-wrap">
+          <code>{logs}</code>
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/modals/LogsModal.tsx]: Btw what's the order of the logs?

**Comment by MrFlashAccount** [comment]
Small question, do we limit the amount of logs to show simultaneously?

**Comment by somebody1234** [comment]
> Small question, do we limit the amount of logs to show simultaneously?

for the time being no

**Comment by MrFlashAccount** [comment]
> > Small question, do we limit the amount of logs to show simultaneously?
> 
> for the time being no

What'll happen if a user has more than, lets say 1 000 000 entries(50mb) of logs? Are we going to process and display all of them?

**Comment by somebody1234** [comment]
@MrFlashAccount we're assuming that won't happen for the time being - we definitely want pagination and scrollback at some point, *however* i think it's not something we should be adding as an afterthought on a random PR

File: app/ide-desktop/lib/dashboard/src/modals/LogsModal.tsx (hunk 3 above)
Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/modals/LogsModal.tsx]: it's a single string here. but i the endpoint returns the logs from oldest to newest (i.e. the same orer as they would appear in a terminal)

**Comment by somebody1234** [comment]
code review should be addressed - not 100% sure the dialog still works properly, *but* we can't test it unless the logs PR is deployed

File: app/ide-desktop/lib/dashboard/src/modals/LogsModal.tsx (hunk 3 above)
Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/modals/LogsModal.tsx]: can we reverse scrolling then? I think showing the oldest logs isn't what users want(often they're looking for the recent error). We can either use bottom-to-top scrolling(using col-reverse order hack in flex) or stream logs in reverse order from lambda

**Comment by somebody1234** [comment]
ping @PabloBuchu for QA?

File: app/ide-desktop/lib/dashboard/src/layouts/AssetPanel.tsx [hunk 4]
```diff
@@ -114,7 +121,7 @@ export default function AssetPanel(props: AssetPanelProps) {
         event.stopPropagation()
       }}
     >
-      <div className="flex">
+      <div className="flex gap-buttons">
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/layouts/AssetPanel.tsx]: please don't use obscure classes like `gap-buttons`.

File: app/ide-desktop/lib/dashboard/src/layouts/AssetProjectSession.tsx [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+          {getText('onDateX', dateTime.formatDateTime(new Date(projectSession.createdAt)))}
+        </time>
+      </div>
+      <div className="flex items-center gap-1">
+        <Button
+          active
+          image={LogsIcon}
+          alt={getText('showLogs')}
+          onPress={() => {
+            setModal(
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/layouts/AssetProjectSession.tsx]: Can we use `ariaComponents.Dialog` here instead?

Also, I'm not sure we need here `setModal`. I think we're overusing it.

Pattern with `<DialogTrigger />` and `<Dialog />` looks more straightforward here

File: app/ide-desktop/lib/dashboard/src/layouts/AssetProjectSession.tsx [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+              <LogsModal
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/layouts/AssetProjectSession.tsx]: oh no, please use `react-query` and please show the loading state. I think putting loading inside the modal should be sufficient. 

One note: automatic loading state works only with `<Suspence />` and `useSuspenceQuery`

File: app/ide-desktop/lib/dashboard/src/layouts/AssetProjectSessions.tsx [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+export interface AssetProjectSessionsProps {
+  readonly item: AssetTreeNode<backendModule.ProjectAsset>
+}
+
+/** A list of previous versions of an asset. */
+export default function AssetProjectSessions(props: AssetProjectSessionsProps) {
+  const { item } = props
+  const { backend } = backendProvider.useStrictBackend()
+
+  const projectSessions = asyncEffectHooks.useAsyncEffect(
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/layouts/AssetProjectSessions.tsx]: same here: lets use react-query.

File: app/ide-desktop/lib/dashboard/src/layouts/AssetProjectSessions.tsx [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+// === AssetProjectSessions ===
+// ============================
+
+/** Props for a {@link AssetProjectSessions}. */
+export interface AssetProjectSessionsProps {
+  readonly item: AssetTreeNode<backendModule.ProjectAsset>
+}
+
+/** A list of previous versions of an asset. */
+export default function AssetProjectSessions(props: AssetProjectSessionsProps) {
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/layouts/AssetProjectSessions.tsx:26]: Please, avoid default exports. They are harmful: https://dev.to/phuocng/avoid-using-default-exports-a1c

File: app/ide-desktop/lib/dashboard/src/modals/LogsModal.tsx [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+  const logsQuery = reactQuery.useQuery({
+    queryKey: ['logs', uniqueString.uniqueString()],
+    queryFn: () => logsRaw,
+  })
+
+  return (
+    <ariaComponents.Dialog type="fullscreen" className="bg-dim">
+      <div
+        tabIndex={-1}
+        className="pointer-events-auto relative flex h-4/5 w-4/5 flex-col gap-4 overflow-auto rounded-default p-4 before:absolute before:inset before:h-full before:w-full before:rounded-default before:bg-selected-frame before:backdrop-blur-3xl"
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/modals/LogsModal.tsx]: I think we can safely remove this element

File: app/ide-desktop/lib/dashboard/src/layouts/AssetProjectSession.tsx [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+                logs={backend
+                  .getProjectSessionLogs(projectSession.projectSessionId, project.title)
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/layouts/AssetProjectSession.tsx]: I am not sure it's a good idea to pass a `promise` instead of `() => Promise` here since we create a fetch request on each render.

File: app/ide-desktop/lib/dashboard/src/modals/LogsModal.tsx [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+        onClick={event => {
+          event.stopPropagation()
+        }}
+      >
+        <aria.Heading level={2} className="relative text-lg font-bold">
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/modals/LogsModal.tsx]: please use `title` prop on dialog instead

File: app/ide-desktop/lib/dashboard/src/layouts/AssetProjectSessions.tsx (hunk 8 above)
Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/layouts/AssetProjectSessions.tsx:26]: addressing each point individually:
- poor discovery: not true, autocomplete picks up default exports with the correct name
  - developers can easily see the exported members of a module: a non-issue. generally react components only export a single component.
- refactoring: yes, this *is* an issue
- auto-complete: as above, works out of the box
- inconsistent codebases: technically true. however:
  - named imports can also trivially be renamed
  - again - auto-imports make this a non-issue in practice.
- re-exporting: only relevant when it is a library. also they missed out on:
```ts
import add from './add';
export { add };
```

tl;dr: all but one of the points simply do not hold up - many of the comments also mirror this sentiment.

even the React official docs prefer default exports: <https://react.dev/learn/referencing-values-with-refs#example-building-a-stopwatch>

so in this case i'd rather follow convention than switch conventions completely for little to no advantage

Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/layouts/AssetProjectSessions.tsx:26]: > named imports can also trivially be renamed

It's not an issue since it's explicit.

> re-exporting: only relevant when it is a library

True, but we use reexporting a lot already

-- 

In the current codebase default export looks more like a bypass of the `import *` rule, don't you think?

this inconsistency leads to a problem when I don't know how to import a function/component upfront - do I need to import default? or should I need to use import * ?

> even the React official docs prefer default exports

They never said that they __prefer__ default exports, see: https://react.dev/learn/importing-and-exporting-components#exporting-and-importing-a-component

CleanShot 2024-06-12 at 15 43 24@2x


--- 

All in all, in some cases, default exports make sense, but due to their `anonymous` and `implicit` nature, they're often forbidden. Too easy to make a mistake.

File: app/ide-desktop/lib/dashboard/src/modals/ProjectLogsModal.tsx [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+      return logs.join('\n')
+    },
+  })
+
+  return (
+    <ariaComponents.Dialog
+      onOpenChange={setIsOpen}
+      title={getText('logs')}
+      type="fullscreen"
+      className="bg-dim"
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/modals/ProjectLogsModal.tsx]: (You can ignore it) we don't need `bg-dim` here, because bg is set properly by the dialog

File: app/ide-desktop/lib/dashboard/src/modals/ProjectLogsModal.tsx [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+/** A modal for showing logs for a project. */
+export default function ProjectLogsModal(props: ProjectLogsModalProps) {
+  const { projectSessionId, projectTitle } = props
+  const { backend } = backendProvider.useStrictBackend()
+  const { getText } = textProvider.useText()
+  const [isOpen, setIsOpen] = React.useState(false)
+  const logsQuery = reactQuery.useQuery({
+    enabled: isOpen,
+    queryKey: ['projectLogs', { projectSessionId, projectTitle }],
+    queryFn: async () => {
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/modals/ProjectLogsModal.tsx:49]: if you're using `useQuery`, then please provide loading and error states, otherwise, put the loading under the `<Dialog />` component so it will handle it for you.

File: app/ide-desktop/lib/dashboard/src/layouts/AssetProjectSessions.tsx [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+    <div className="pointer-events-auto flex flex-col items-center overflow-y-auto overflow-x-hidden">
+      {projectSessionsQuery.isSuccess ? (
+        projectSessionsQuery.data.map(session => (
+          <AssetProjectSession
+            key={session.projectSessionId}
+            project={item.item}
+            projectSession={session}
+          />
+        ))
+      ) : projectSessionsQuery.isLoading || projectSessionsQuery.isPending ? (
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/layouts/AssetProjectSessions.tsx]: is loading should be enough.

We can flip the order and render data after loading / error states and thus we don't need to check if query is success

File: app/ide-desktop/lib/dashboard/src/modals/ProjectLogsModal.tsx [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+  readonly projectSessionId: backendModule.ProjectSessionId
+  readonly projectTitle: string
+}
+
+/** A modal for showing logs for a project. */
+export default function ProjectLogsModal(props: ProjectLogsModalProps) {
+  const { projectSessionId, projectTitle } = props
+  const { backend } = backendProvider.useStrictBackend()
+  const { getText } = textProvider.useText()
+  const [isOpen, setIsOpen] = React.useState(false)
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/modals/ProjectLogsModal.tsx]: You don't need this state here. use `<DialogTrigger />`

File: app/ide-desktop/lib/dashboard/src/modals/ProjectLogsModal.tsx [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+  const logsQuery = reactQuery.useQuery({
+    enabled: isOpen,
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/modals/ProjectLogsModal.tsx]: We can safely remove this as well

File: app/ide-desktop/lib/dashboard/src/layouts/AssetPanel.tsx [hunk 17]
```diff
@@ -134,6 +141,25 @@ export default function AssetPanel(props: AssetPanelProps) {
               {getText('versions')}
             </ariaComponents.Button>
           )}
+        {item != null && item.item.type === backendModule.AssetType.project && (
+          <ariaComponents.Button
+            size="medium"
+            variant="ghost"
+            className={tailwindMerge.twMerge(
+              'button pointer-events-auto select-none bg-frame px-button-x leading-cozy transition-colors hover:bg-selected-frame',
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/layouts/AssetPanel.tsx]: oh what's that? why do you provide that much of custom classes together with `variant` prop?

File: app/ide-desktop/lib/dashboard/src/layouts/AssetProjectSessions.tsx (hunk 8 above)
Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/layouts/AssetProjectSessions.tsx:26]: > In the current codebase default export looks more like a bypass of the `import *` rule, don't you think?

no, as mentioned above it is because of React conventions. if not for that, i would rather namespace all components rather than using default exports...

> they never say that they *prefer* default exports.

true. however, *every single example in their docs* uses default exports. (the same applies to next.js docs.)

while it's true that export style tends to be divided mostly evenly between default and named exports, it's most certainly not a valid reason why named exports are better, or why default exports are worse.

> anonymous nature

not true, they take the names of the variable that is exported. it is *near impossible* to make a mistake because they, like everything else, are typically imported using auto-imports.

it's true that this may cause issues with things being out of sync when *renaming a file*. however this is an extremely rare occurrence and so I don't think it matters a lot

> we use reexporting a lot already

IMO re-exporting is mostly a non-concern when the component is only ever used in one place - i'm not convinced that they improve the organization of the code, not when the directory structure already offers the exact same level of organization

> this inconsistency leads to a problem when I don't know how to import a function/component upfront - do I need to import default? or should I need to use import * ?

public components were always default exports, so *this was a non-issue* until named exports were introduced - we have used default exports ever since the very beginning: <https://github.com/enso-org/enso/blob/6f29262f90027995efbe8d5eee6d831496e97ad7/app/ide-desktop/lib/dashboard/src/authentication/src/dashboard/components/dashboard.tsx#L21>

Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/layouts/AssetProjectSessions.tsx:26]: one final note is that it's worth noting that Vue also prefers default exports: <https://github.com/enso-org/enso/blob/0e17beba73884f06f352b92e20312712df7afdc5/app/gui2/src/components/GraphEditor/GraphNodes.vue#L2-L3>

**Comment by PabloBuchu** [comment]
1. When I open asset panel and switch from versions to session the whole dashboard (I see logging to enso) is reloading and then the session is being loaded


https://github.com/enso-org/enso/assets/2855109/1ebef0fe-8ed2-493c-81b9-ee641dc207c1



2. Entries on the list looks like they are disabled
Screenshot 2024-06-18 at 12 10 19

**Comment by somebody1234** [comment]
yeah my bad, couldn't test the new changes without the lambdas being accessible, will test soon™

**Comment by somebody1234** [comment]
@PabloBuchu QA issues should be addressed now.

not sure if the session ids are ideal (since they are just the times currently)... for now should we just name them by index like we currently do with versions?

**Comment by PabloBuchu** [comment]
@somebody1234 can you tomorrow merge current develop? I will be able to reqa then

**Review by PabloBuchu — APPROVED** [review]

**Review by MrFlashAccount — APPROVED** [review]

File: app/ide-desktop/lib/dashboard/src/components/AriaComponents/Dialog/types.ts [hunk 18]
```diff
@@ -5,6 +5,7 @@ import type * as aria from '#/components/aria'
 export interface DialogProps extends aria.DialogProps {
   readonly title?: string
   readonly isDismissable?: boolean
+  readonly isOpen?: boolean
```
Comment (MrFlashAccount) [inline app/ide-desktop/lib/dashboard/src/components/AriaComponents/Dialog/types.ts]: @somebody1234 we have `modalProps` for that

Comment (MrFlashAccount) [reply app/ide-desktop/lib/dashboard/src/components/AriaComponents/Dialog/types.ts]: I guess we can flatten `modalProps` because it seems like it's not clear that we can control the open/close using it

Comment (somebody1234) [reply app/ide-desktop/lib/dashboard/src/components/AriaComponents/Dialog/types.ts]: just tried, seems like `className` conflicts with `aria.DialogProps` which this interface also extends. but it seems like it's not being used anyway so i've just removed it
